### DISCUSSION TITLE
Rationale for path canonicalize?

### DISCUSSION THREAD

**Discussion by tesujimath** [nushell/nushell#12628] [General]
I am trying to understand why Nu does what it does with regard to aggressive canonicalization of paths.  I have hunted around various issues and found some snippets, but nowhere did I find a nicely reasoned rationale for this is what Nu does and why.  Is there such a thing?

Fully acknowledging that I am coming to this quite late in the piece, my feeling is that I am in the early stages of a long and rewarding relationship with Nu, and so I am wanting to engage in this discussion so that in the long term I have a sense that Nu's handling of paths is conceptually clean.

My perspective is that aggressive canonicalization is conceptually misguided.  In particular, the readiness with which Nu resolves symlinks into their canonical paths is quite uncomfortable, and at times, even painful.  But rather than simply stating a contrary opinion, I would like to give a handful of examples, and then state my general principle here, with a view to eliciting a helpful discussion.

1. [Nix](https://nixos.org/) makes heavy use of symlinks, with for example symlinks from a user's profile directory (under their home) into scary looking paths in the Nix store. (I fully understand that not everyone is interested in Nix.  Please just accept for now that it is truly wonderful, and is in fact my rest-of-life Linux distro.)

```
> ls -l (which nu | get 0 | get path) | select name type target
╭───┬───────────────────────────────┬─────────┬───────────────────────────────────────────────────────────────────╮
│ # │             name              │  type   │                              target                               │
├───┼───────────────────────────────┼─────────┼───────────────────────────────────────────────────────────────────┤
│ 0 │ /home/sjg/.nix-profile/bin/nu │ symlink │ /nix/store/rifvkpcbf4z45i0ask5qiiip5xn4sh7l-nushell-0.91.0/bin/nu │
╰───┴───────────────────────────────┴─────────┴───────────────────────────────────────────────────────────────────╯
```

With Nix Home Manager we find this sort of thing very frequently occuring:
```
> ls -l ~/.bashrc | select name type target
╭───┬───────────────────┬─────────┬────────────────────────────────────────────────────────────────────────╮
│ # │       name        │  type   │                                 target                                 │
├───┼───────────────────┼─────────┼────────────────────────────────────────────────────────────────────────┤
│ 0 │ /home/sjg/.bashrc │ symlink │ /nix/store/12jv1ny9rfzgy7nz5gv2fm5ayiyp9z9h-home-manager-files/.bashrc │
╰───┴───────────────────┴─────────┴────────────────────────────────────────────────────────────────────────╯

> ls -l ~/.config/nushell/plugins/nu_plugin_bash_env | select name type target
╭───┬──────────────────────────────────────────────────────┬─────────┬───────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ # │                         name                         │  type   │                                                  target                                                   │
├───┼──────────────────────────────────────────────────────┼─────────┼───────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ 0 │ /home/sjg/.config/nushell/plugins/nu_plugin_bash_env │ symlink │ /nix/store/12jv1ny9rfzgy7nz5gv2fm5ayiyp9z9h-home-manager-files/.config/nushell/plugins/nu_plugin_bash_env │
╰───┴──────────────────────────────────────────────────────┴─────────┴───────────────────────────────────────────────────────────────────────────────────────────────────────────╯
```

2. [GNU Stow](https://www.gnu.org/software/stow/) is a useful program for installing unrelated directory trees into a common place (e.g. `/usr/local`) by means of a link farm.  Files which end up appearing to live in the same directory are in fact nowhere near each other when symlinks are resolved.

3. Bash has a long-established way of maintaining a logical working directory distinct from the physical one.  So users coming from Bash have a certain expectation, for which Nu's behaviour is a bit of a gotcha.

```
> cd ~/junk
> pwd
/home/sjg/junk
> ls -l d1 | select name type target
╭───┬───────┬─────────┬────────╮
│ # │ name  │  type   │ target │
├───┼───────┼─────────┼────────┤
│ 0 │ d1/d2 │ symlink │ ../d2  │
╰───┴───────┴─────────┴────────╯

> cd d1/d2
> pwd
/home/sjg/junk/d2
> cd ..
> pwd
/home/sjg/junk

> bash
$ pwd
/home/sjg/junk
$ cd d1/d2
$ pwd                    # bash builtin, logical path
/home/sjg/junk/d1/d2
$ /usr/bin/env pwd       # physical path
/home/sjg/junk/d2
$ cd ..
$ pwd
/home/sjg/junk/d1
```

So here's my principle on this.  I think somehow these symlinks provide the "public interface" if you like to these paths.  Resolving symlinks is rather like violating this encapsulation, and exposing internal details (physical paths) where this is not really appropriate.

Is there a good reason for canonicalizing paths in some cases?  Has it perhaps been applied a little too universally?

**Comment by ChrisDenton** [comment]
There's a related PR that does change nu's behaviour https://github.com/nushell/nushell/pull/12515

**Reply by tesujimath** [reply]
Thanks @ChrisDenton, that PR is evidence that aggressive canonicalization is not a panacea.  I think more work in this whole area is required if Nu's behaviour is to be correct and intuitive.  I'm pleased to see that PR anyway.

**Comment by fdncred** [comment]
I could totally be making this up due to remembering wrong, but I remember starting to do some canonicalization stuff for 2 reasons.
1. We were having issues sometimes finding config files.
2. We wanted the correct path locations to show up when you execute `$nu`.

As I recall, there are a handful of issues of nushell not finding files properly, the most recent that I remember happened when we added XDG_CONFIG_HOME support.

**Comment by tesujimath** [comment]
Thanks @fdncred. Manipulating paths containing symlinks is clearly tricky, and I can see why Nu has taken the solution of canonicalize everything in general.

#12515 shows that this is not desirable in general.  As do my own examples above.

This must be a solved problem. though - it's not as if Nu is the first shell to face this!  I am inclined to look into this some more.

**Reply by fdncred** [reply]
I saw someone say that we may have done this to also help with problems with `rm`, `cp`, `mv`, `open`, and other file i/o commands.

**Comment by tesujimath** [comment]
I think it will be worth listing the cases where canonicalize has been used to solve real or potential breakage.  Let me start.

1. Given a path to a file, open another file relative to that.  For example, a link to config file which includes another (expected perhaps to be in the same directory).

**Comment by tesujimath** [comment]
And separately, potential solutions to these problem cases, which preserve the desired user view of non-expanded symlinks.

To open a file relative to another file, we could envisage a `file_directory()` function, which essentially is a late-bound canonicalization of its path, followed immediately by taking the directory.

In contrast, what Nu currently does could be called early bound canonicalization.

I am gently arguing in favour of late bound canonicalization, for reasons of increased usability and greater intuition.  But this is early days in this line of reasoning.

**Comment by RossSmyth** [comment]
Yeah in general I think canonicalization is something an application should not do unless it is forced to.
